### ISSUE TITLE
Proposal for WebGL rendering

### ISSUE THREAD

**Issue by thejohnhoffer** [openseadragon/openseadragon#1482]
I'm working with @jmuhlich and @dpwrussell on an application using OpenSeadragon. Along the way, we've outlined one approach to a WebGL rendering engine within OpenSeadragon. We'd like to know whether these changes would be of general interest, and we're looking for feedback on the general approach outlined below.

---

We propose changes to OpenSeadragon to replace canvas 2D rendering with rendering through WebGL. The proposed OpenSeadragon updates will allow application-specific “GLSL” fragment shaders to modify all pixels in tiled images. Custom shaders will let applications render color images directly from a variety of useful scientific encodings such as 2D label maps or high dynamic range grayscale images.

All browsers supporting OpenSeadragon’s use of canvas also support the proposed use of WebGL. No up-to-date browsers will benefit from a 2D canvas fallback from WebGL; therefore, we propose removing “2d” contexts entirely to reduce the number of implementations to maintain. After feature detection, OpenSeadragon will use either a “webgl2” context, a “webgl” context, or the HTML rendering fallback.

If OpenSeadragon detects WebGL support, each image record will manage `TypedArray` objects from decoded ajax responses or from rendered image elements. Methods of the drawer, tiled images, and tiles will render these arrays to textures and bind all textures per frame to a webgl context. Without WebGL, the image record will still manage image elements directly. In either case, no rendered 2D canvas context will be created or stored by the image job. With WebGL, the drawer will call a new method of Tile named `drawGL`, which will use a cached TypedArray. Otherwise, the drawer will still call `drawHTML` from a cached image element. In either case, the `drawCanvas` method will not be called, and the sketch canvas will not be needed.

We will add to the API, updating the interfaces and descriptions in the OpenSeadragon docs. The API will remain compatible with the current (version 2.3.x) API; however, plugins will break if they rely on overwriting methods that Openseadragon will no longer call internally. New external API endpoints will allow projects depending on OpenSeadragon to supply custom functions written in GLSL ES 1.00 to use in all current browsers and/or GLSL ES 3.00 to use in browsers with detected webgl2 support.

### Open Questions:

- With Ajax enabled and WebGL supported, we intend to store pixel data in `TypedArray`s instead of HTML image elements to support high dynamic range grayscale images common in scientific imaging. As HTML image elements only provide 8-bit pixel data from standard HTML5 image formats, we will need to support loading from higher-depth formats through JavaScript implementations of PNG (16-bit), TIFF, etc. We are unsure whether to bundle these decoders within OpenSeadragon or ask applications to pass a custom decoder as a tilesource option.

- OpenSeadragon will cease to use certain documented properties and methods. For example, the viewer event `tile-loaded` supplies an HTML image element. Even when no image element is needed internally, the new implementation of such events must always supply an image element and handle modifications to it in order to maintain compatibility with existing plugins. The documentation only references some properties (e.g. a drawer’s `SketchCanvas`) indirectly within descriptions of other properties and methods. We are unsure whether to remove such properties entirely.

- We suggest continuing to handle the pyramids generated by an `ImageTileSource` by initializing tiles with “2d” contexts, but these contexts will be used to either call `drawGL` on a `TypedArray` or `drawHTML` on a HTML image element. We are unsure whether any simpler solution could better support the API.

**Comment by iangilman** [comment]
Exciting! We've been interested in something like this for a long time (see #68 for webGL and #1380 and #1215 for image format support). Just to be clear, I believe the advantage we're seeking with webGL is performance. 

The addition of support for other formats, including those with deeper bit depth, also sounds great, though I suppose that change would still be possible with canvas (though perhaps it would require too much pixel manipulation to be performant). 

I don't think I can support removing the canvas draw path immediately, however. I think in order to make this a successful transition we're going to need to maintain both the canvas and webGL draw paths for at least some period of time. For one thing, as you point out, many plugins rely on the canvas draw path, and it would be good to give them a chance to migrate to the webGL draw path before breaking them. This can only happen if there is a period of overlap where both options are available.

I know from my own development experience on other projects that while all of the major browsers currently support webGL, they don't all do it better than canvas. IE 11 has a number of bugs related to webGL, and lower powered mobile devices also have issues with webGL compared to canvas. For this reason, webGL isn't an automatic win for every device/browser, so we'd be doing a disservice to simply move everyone over.

So, I propose moving forward with a new webGL draw path that can coexist with the canvas draw path. Perhaps to cut down on the initial complexity, that can happen at first with regular images and the additional image format support can be added once the new draw path is in place. 

Regarding the new formats, I'll be interested to know the trade offs involved with internal support vs. external. I see three options for each format: Supported directly within OSD, supported via plugin, or supported via user code. Given the different trade offs for different formats, I can imagine some formats would be better with one form of support where as others would be better with another. I'd love to see a general purpose "format support" mechanism.

Like I said, I'm excited; both webGL support and new format support are big features that have been of interest for a long time!

**Comment by thejohnhoffer** [comment]
That's great! Thanks for your support.

To some extent support for images with higher bit depths is a separate issue from a webGL drawer, but a solution without webGL would require _something_ to reformat the images as 8-bit RGBA for canvas. According to the MDN documentation ,`ctx. putImageData ` and `ctx.drawImage` support `ImageData` and `CanvasImageSource `, respectively. Certainly `ImageData` and apparently all types in `CanvasImageSource` only support 8-bit RGBA. There may be a method I haven't considered.

The format support interface could work similarly to the `TileSource` interface. There are a few built-in options, but then there's one `getTileUrl` function that's easy to understand but generally useful for user-defined code. It would require careful thought to come up with a `getImagePixels` interface that is similarly useful, but it would be worth the effort.

 In terms of compatibility, it's good to know about the WebGL bugs in IE 11 or mobile devices. I agree that maintaining canvas rendering is the easiest way to guarantee backwards-compatible plugin support. I'm still uncertain how to handle plugins that override canvas methods that will only be called when WebGL is not enabled. For that reason, do you think WebGL be enabled with a custom flag, or perhaps enabled by default except in specific browsers?

**Comment by thejohnhoffer** [comment]
If a `getImagePixels` entry point:
  - accepts `ArrayBuffer` from ajax response
  - returns a `TypedArray`
The "2d" canvas draw path should only allow `getImagePixels` to return:
  - `UInt8Array`
The "webgl" draw path should only allow `getImagePixels` to return:
  - `UInt8Array`
  - `UInt16Array`
The "webgl2" draw path should allow `getImagePixels` to return:
  - `UInt8Array`
  - `UInt16Array`
  - `UInt32Array`
  - `Float32Array`

**Comment by iangilman** [comment]
I'm imagining we'll want the default to be an auto-detect that uses WebGL on most browsers but chooses canvas for some. Then we can have an option that forces it to one or the other if the developer needs.

We'll probably want to add an API that can tell you which drawer is being used, so moving forward, plugins that require one or the other can gracefully disable themselves as needed. I could also imagine plugins that would support both, but would need to know which thing to hook into. 

We might also want to support things like a canvas based overlay; no reason we couldn't have a 2d canvas over top of a WebGL canvas.

Presumably if we wanted to "support" (i.e. not break entirely) 16- and 32-bit images on canvas we could crunch them down to 8-bit (along with a warning that that's what's happening). 

BTW, while we're in this area, the ability to generate and/or customize tiles after they've been loaded but before they've been drawn is just kind of hacked in to the system. It would be great if that stuff could be cleaned up a bit in the process. Just a thought.

Anyway, figuring out how to break this down into sensible chunks is probably the best way to keep us all sane! :)

**Comment by thejohnhoffer** [comment]
@iangilman Thank you for your help planning this. I'm going to start writing a standalone drawer. After I have something worth sharing, I'll return to your notes here to begin a more detailed API specification.

I think it seems sensible to break the project up along the various API endpoints needed. While I'm not particularly attached to any of these details, here are my initial thoughts on each endpoint:

- Custom tile generation:
  - Developer defines `loadResponseArray`, given the URL:
      - It functions like `makeAjaxRequest`
      - It returns a `UInt8Array` that is arbitrarily encoded
      - Default to `makeAjaxRequest` or html `image` loading
  - Developer defines `readResponseArray`, given the response:
      - It can decode, unpack, or otherwise modify the bit 
      - It returns a `TypedArray` with actual image pixels
      - Default to decoding jpg/png/gif (as html `image` does)
        - Log warning when reducing higher bit-depths

- Custom WebGL code:
  - Developer requests a type (e.g. `vec2`) for tile pixels
  - Developer defines a fragment shader to act on tile pixels
      - It can be glsl 1.0 (for WebGL 1 or 2) or glsl 3 for (WebGL 2 only):
      - It must set a `vec4` called `rgba`
      - It can read these values
         - Viewport-coordinate bounds of tiled image
         - Image-coordinate bounds of tile
         - World index of tiled image
         - Sampler `u_tile` giving tile (as requested type)
         - A`uv` holding `vec2` tile coordinates

- Which drawer to use
   - Viewer option `useWebGL: true` overrides `useCanvas: true`
   - Developers can read `Drawer.useWebGL` xor `Drawer.useCanvas` 
     - Or `Drawer.contextType` of [undefined, "2d", "webgl", webgl2]
   - Feature detection
     - WebGL 2: anything with `getContext('webgl2')`
     - WebGL: anything with `getContext('webgl')`
       - except ie11, old mobile browsers
     - Canvas 2D: anything with `getContext('2d')`

**Comment by iangilman** [comment]
Looks great to me! And yes, starting with the drawer seems like a good plan. I'm excited to see this in action!

**Comment by yellowtailfan** [comment]
I am considering coding something similar in purpose to what @thejohnhoffer has proposed. I have similar needs:
* View multi-channel high bit depth (uint16, float16, or float32) image data
* Mapping from multi-channel high bit depth to 8-bit RGB for display
* User interface for making interactive adjustments (e.g. gamma, offset, gain, range, normalise) to the mapping quickly on the client side without re-fetching image data from the server

Openseadragon and viaWebGL look like excellent starting points for this.

I have a concern about the performance of using pixelwise loops in JavaScript for data conversions. I found an example from a LeafletsJS plugin which [does this kind of pixelwise data conversion in GLSL](https://github.com/ihmeuw/leaflet.tilelayer.glcolorscale/blob/master/src/shaders/util/rgbaToFloat.glsl).

Are any status updates on the work proposed in this issue? It would be good to know before I start coding.

**Comment by jmuhlich** [comment]
Great to have your support, @yellowtailfan ! @thejohnhoffer had to take a break so our team hasn't had a chance to finish the specification or write any code, but we're still desperate for this feature in OSD. Maybe together we can flesh out the API spec outlined above, before diving in to the coding? You can find my email in my profile and contact me to work out a plan. I'm also curious what you're working on! (We're visualizing many-channel fluorescence microscopy images)

**Comment by yellowtailfan** [comment]
Thanks @jmuhlich, that sounds great. I’ll be in touch soon.

My application is related: I’m working on scientific image processing. 

I also found another approach to high bit depth data in the [JavaScript Extended-Range Image (JERI) Viewer](https://jeri.io/). They used emscripten to compile the OpenEXR C/C++ library into WebAssembly. OpenEXR is a multi channel high bit depth image file format used for visual effects (JERI was written by Disney Research). 

I don’t know much about performance optimisation in modern JavaScript. Can anyone comment on the relative performance of pixel-wise data type conversion and range mapping in:
* Pure JavaScript
* WebAssembly
* WebGL shaders

I would guess that WebGL shaders would be the fastest, but I don’t know how much difference there would be between the 3 options. I imagine there would be browser compatibility issues too, especially on mobile.

**Comment by iangilman** [comment]
I would guess WebGL shaders as well, since they are run straight on the GPU just like shaders that come from native code. I don't have any hard data to back that up. Seems like some experimentation/benchmarking might be in order.

**Comment by iangilman** [comment]
Possibly relevant to this issue, @karin-toth posted in https://github.com/openseadragon/openseadragon/issues/1683#issuecomment-574162070 a link to her stab at a WebGL renderer for OpenSeadragon:

https://github.com/karin-toth/openseadragon/tree/kato-webgl

Her comments:

> I've been working on a webGL renderer for quite some time now, but as many other things, it's more complex that it seems. I got something to work, but taking it from a seems to work-state to the type of fail-safe production code that is required takes more time than I thought. So I would have loved to contribute, but I just don't have the time to complete this.
> 
> I'm of course open for answering questions and other help in the future if someone wants to continue.

**Comment by maxbogue** [comment]
Any updates on this from the maintainers? If OSD has anything akin to a roadmap, is this on it? Or will someone need to contribute a fully polished PR for it to happen?

**Comment by iangilman** [comment]
Nothing new on this front. All of the work on OSD is from volunteers, so we don't have a roadmap... It's just whatever someone is motivated to work on. I feel like there's potential for the many institutions that use OSD to kick in some money for more directed development (and we do have [an Open Collective](https://opencollective.com/openseadragon)), but I think in order to make that happen, we'll need someone who can actually go around and hit those institutions up. I'm thinking it may be time to start looking for that person, so if you (or anyone else) have suggestions, I'm all ears! 

But I digress. No, nothing new on the WebGL front. Every now and then someone gets interested and takes a stab at it, but none of it has come to fruition yet. A company almost hired me to do it a year or two ago, but that didn't end up happening either. I think I've got a pretty good idea of how to make it happen, but it's a big job and I can't afford to take that much time off from work without getting paid for it. 

If you're interested in tackling it, or if you know someone who might want to fund it, dive in! I'm happy to help point the way :)

**Comment by iangilman** [comment]
This is now superseded by #2310.
